src/ln/channelmonitor.rs

```
@@ -516,38 +622,81 @@ impl ChannelMonitor {
-				return Err(MonitorUpdateError("Funding transaction outputs are not identical!"));
-			}
-		} else {
-			self.funding_txo = other.funding_txo.take();
-		}
+
+		self.key_storage = match self.key_storage {
+			Storage::Local { ref revocation_base_key, ref htlc_base_key, ref delayed_payment_base_key, ref payment_base_key, ref shutdown_pubkey, ref prev_latest_per_commitment_point, ref latest_per_commitment_point, ref mut funding_info, ref mut short_channel_id, .. } => {
+
+				macro_rules! new_storage_local {
+					($funding_info: expr, $short_channel_id: expr) => {
+						Storage::Local {
+							revocation_base_key: *revocation_base_key,
+							htlc_base_key: *htlc_base_key,
+							delayed_payment_base_key: *delayed_payment_base_key,
+							payment_base_key: *payment_base_key,
+							shutdown_pubkey: *shutdown_pubkey,
+							prev_latest_per_commitment_point: *prev_latest_per_commitment_point,
+							latest_per_commitment_point: *latest_per_commitment_point,
+							funding_info: $funding_info,
+							short_channel_id: $short_channel_id,
+						}
+					}
+				}
+
+				let our_funding_info = funding_info;
+				let our_short_channel_id = short_channel_id;
+				if let Storage::Local { ref mut funding_info, ref mut short_channel_id, .. } = other.key_storage {
+					if our_funding_info.is_some() {
+					// We should be able to compare the entire funding_txo, but in fuzztarget its trivially
+					// easy to collide the funding_txo hash and have a different scriptPubKey.
+						if funding_info.is_some() && our_funding_info.is_some() && funding_info.as_ref().unwrap().0 != our_funding_info.as_ref().unwrap().0 {
+							return Err(MonitorUpdateError("Funding transaction outputs are not identical!"));
+						} else {
+							new_storage_local!(our_funding_info.take(), our_short_channel_id.take())
+						}
+					} else {
+						new_storage_local!(funding_info.take(), short_channel_id.take())
```

Hmm? I'm still really not convinced we need to support funding_info.is_none() in most places in ChannelMonitor. Specifically if I replace this line with a panic!() the tests still pass, and I don't think there *should* be a way to get to this anyway (it should return an Err as otherwise you could crash a watchtower, but the tests still pass fine without it, which could simplify this function).

---

Yes, replaced by Err, in the current state, it's technically possible to hit this branch if user pass us a ChannelMonitor without funding_info set but agree it wouldn't make sense and not hit (I think) but our current code flow. Should go away with refactor in Channel as previously talked